### PR TITLE
Save storage data in snapshot

### PR DESCRIPTION
**Pull request by e-schepachev**
-

### CODE REVIEW

File: plugins/producer_api_plugin/producer_api_plugin.cpp [hunk 1]
```diff
@@ -81,6 +81,8 @@ void producer_api_plugin::plugin_startup() {
 //            INVOKE_R_V(producer, get_integrity_hash), 201),
 //       CALL(producer, producer, create_snapshot,
 //            INVOKE_R_V(producer, create_snapshot), 201),
```
Comment (s-medvedev) [inline plugins/producer_api_plugin/producer_api_plugin.cpp]: As I can see this line simply uncommented. Please remove it.

File: plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp [hunk 2]
```diff
@@ -79,3 +79,4 @@ FC_REFLECT(eosio::producer_plugin::runtime_options,
 //TODO: Removed by CyberWay
 //FC_REFLECT(eosio::producer_plugin::integrity_hash_information, (head_block_id)(integrity_hash))
 //FC_REFLECT(eosio::producer_plugin::snapshot_information, (head_block_id)(snapshot_name))
```
Comment (s-medvedev) [inline plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp]: Please remove this line.

File: libraries/chain/controller.cpp [hunk 3]
```diff
@@ -531,22 +531,137 @@ struct controller_impl {
...
+    }
+
+    void dump_object(snapshot_writer::section_writer& section, const cyberway::chaindb::object_value& object) const {
+        const auto value = object.value;
+        const auto service = object.service;
 
-      set_revision( head->block_num );
+        fc::variant service_variant;
+        fc::to_variant(service, service_variant);
+        const fc::variant full_object = fc::mutable_variant_object(value.get_object())("SERVICE", service_variant);
```
Comment (s-medvedev) [inline libraries/chain/controller.cpp]: I think `_SERVICE_` will be better for service field.

File: libraries/chain/controller.cpp [hunk 4]
```diff
@@ -531,22 +531,137 @@ struct controller_impl {
...
+        for (const auto& abi : abis) {
+           dump_contract(snapshot, abi);
+        }
+    }
 
-   void read_from_snapshot( const snapshot_reader_ptr& snapshot ) {
-      // TODO: Removed by CyberWay
+    std::vector<cyberway::chaindb::abi_info> dump_accounts(const snapshot_writer_ptr& snapshot) const {
+        std::vector<cyberway::chaindb::abi_info> abis;
+        snapshot->write_section("account_table", [&] (auto& section) {
```
Comment (s-medvedev) [inline libraries/chain/controller.cpp]: This section may be overwritten by `table` table from `account` contract.

File: libraries/chain/controller.cpp [hunk 5]
```diff
@@ -531,22 +531,137 @@ struct controller_impl {
...
+                if (!account.abi.empty()) {
+                    abis.emplace_back(account.name, account.get_abi());
+                }
+                section.add_row(account);
+            });
+        });
+        return abis;
+    }
+
+    void dump_contract(const snapshot_writer_ptr& snapshot, const cyberway::chaindb::abi_info& abi) const {
```
Comment (s-medvedev) [inline libraries/chain/controller.cpp]: May be rename this function to `dump_contract_tables`?

File: libraries/chain/controller.cpp (hunk 4 above)
Comment (e-schepachev) [reply libraries/chain/controller.cpp]: at line 557:
 if (cyberway::chaindb::is_system_code(abi.code()) && table.first == account_table::table_name()) 
this case is filtered out

**Pull request closed without merging**